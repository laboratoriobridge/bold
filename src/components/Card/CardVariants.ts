import { ExternalStyles, focusBoxShadow, hexToRGB, Theme } from '../../styles'

export type CardVariant = 'outline' | 'float' | 'plain'

export const createBaseStyles = (theme: Theme) => ({
  card: {
    width: '100%',
    border: 0,
    borderRadius: `${theme.radius.popper}px`,
    padding: '1rem',
    transition: 'all .2s',
  } as ExternalStyles,
  cardDisabled: {
    cursor: 'not-allowed',
    '& *': {
      pointerEvents: 'none',
    },
  } as ExternalStyles,
  innerDisabled: {
    opacity: 0.4,
  } as ExternalStyles,
})

export const createVariantStyles = (theme: Theme): { [key in CardVariant]: ExternalStyles } => ({
  float: {
    background: theme.pallete.surface.main,
    border: `1px solid ${theme.pallete.gray.c80}`,
    boxShadow: theme.shadows.outer[40],
    '&[data-invalid="true"]': {
      borderColor: theme.pallete.status.danger.main,
    },
    '&[data-disabled="true"]': {
      borderColor: theme.pallete.gray.c80,
      background: theme.pallete.surface.main,
    },
  } as ExternalStyles,
  outline: {
    background: theme.pallete.surface.main,
    border: `1px solid ${theme.pallete.gray.c80}`,
    borderRadius: `${theme.radius.popper}px`,
    '&[data-invalid="true"]': {
      borderColor: theme.pallete.status.danger.main,
    },
    '&[data-disabled="true"]': {
      borderColor: theme.pallete.gray.c80,
      background: theme.pallete.surface.main,
    },
  } as ExternalStyles,
  plain: {
    background: theme.pallete.surface.main,
    borderRadius: `${theme.radius.popper}px`,
    '&[data-invalid="true"]': {
      background: theme.pallete.status.danger.c90,
    },
    '&[data-disabled="true"]': {
      borderColor: theme.pallete.gray.c80,
      background: theme.pallete.surface.main,
    },
  } as ExternalStyles,
})

export const createClickableStyles = (theme: Theme): ExternalStyles => ({
  ...theme.typography.variant('main'),
  cursor: 'pointer',
  textAlign: 'initial',
  '&[data-variant="outline"], &[data-variant="plain"]': {
    borderColor: theme.pallete.gray.c60,
  },
  ':not(:disabled):hover': {
    background: hexToRGB(theme.pallete.gray.c50, 0.1),
    '&[data-variant="float"]': {
      boxShadow: theme.shadows.outer[160],
    },
  },
  ':focus': {
    outline: 'none',
    boxShadow: focusBoxShadow(theme),
    '&[data-variant="float"]': {
      boxShadow: `${theme.shadows.outer[40]}, ${focusBoxShadow(theme)}`,
    },
  },
  ':not(:disabled):not(:active):hover:focus': {
    '&[data-variant="float"]': {
      boxShadow: `${theme.shadows.outer[160]}, ${focusBoxShadow(theme)}`,
    },
  },
  ':not(:disabled):active': {
    boxShadow: theme.shadows.inner['10'],
    '&[data-variant="float"]': {
      // use zero box shadow to keep transition smooth
      boxShadow: `0 0 0 0 rgba(0, 0, 0, 0), 0 0 0 0 rgba(0, 0, 0, 0), 0 0 0 0 rgba(0, 0, 0, 0), ${theme.shadows.inner['10']}`,
    },
  },
  '&[data-selected="true"]': {
    background: theme.pallete.primary.c90,
    borderColor: theme.pallete.primary.main,
    ':not(:disabled):hover': {
      background: theme.pallete.primary.c80,
    },
    '&[data-invalid="true"]': {
      background: theme.pallete.status.danger.c90,
    },
  },
  '&[data-invalid="true"]': {
    borderColor: theme.pallete.status.danger.main,
    ':not(:disabled):hover': {
      background: theme.pallete.status.danger.c80,
    },
    ':focus': {
      outline: 'none',
      boxShadow: focusBoxShadow(theme, 'danger'),
    },
    ':not(:disabled):active': {
      boxShadow: theme.shadows.inner['10'],
    },
    ':not(:disabled):not(:active):hover:focus': {
      '&[data-variant="float"]': {
        boxShadow: `${theme.shadows.outer[160]}, ${focusBoxShadow(theme, 'danger')}`,
      },
    },
  },
  ':disabled': {
    borderColor: theme.pallete.gray.c80,
    '&[data-selected="true"]': {
      borderColor: theme.pallete.primary.c80,
    },
    '&[data-invalid="true"]': {
      borderColor: theme.pallete.gray.c80,
      background: theme.pallete.surface.main,
      '&[data-selected="true"]': {
        borderColor: theme.pallete.status.danger.c80,
        background: theme.pallete.status.danger.c90,
      },
    },
  },
})
