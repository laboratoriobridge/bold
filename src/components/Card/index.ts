export { Card, CardProps } from './Card'
export { ClickableCard, ClickableCardProps } from './ClickableCard'
export { CardVariant } from './cardVariants'
