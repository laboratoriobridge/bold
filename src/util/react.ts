import { MutableRefObject, Ref } from 'react'

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}
