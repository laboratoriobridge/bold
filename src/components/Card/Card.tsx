import React from 'react'

import { css } from 'emotion'
import { ExternalStyles, Theme, useTheme } from '../../styles'
import { CardVariant, createBaseStyles, createVariantStyles } from './CardVariants'

export interface CardProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'style'> {
  style?: ExternalStyles

  /**
   * Defines the visual style of the Card
   * @default 'outline'
   */
  variant?: CardVariant

  /**
   * Specify whether the Card should have error appearance
   */
  invalid?: boolean

  /**
   * Specify whether the Card should have disabled appearance
   */
  disabled?: boolean
}

export function Card(props: CardProps) {
  const { variant = 'outline', children, invalid, disabled, style, ...rest } = props

  const theme = useTheme()
  const styles = createStyles(theme, variant, disabled, invalid, style)

  return (
    <div className={styles.card} {...rest}>
      <div className={styles.inner}>{children}</div>
    </div>
  )
}

const createStyles = (
  theme: Theme,
  variant: CardVariant,
  disabled: boolean,
  invalid: boolean,
  externalStyles: ExternalStyles
) => {
  const baseStyles = createBaseStyles(theme)
  const variantStyles = createVariantStyles(theme)[variant]

  return {
    card: css(
      baseStyles.card,
      variantStyles.base,
      invalid && variantStyles.invalid,
      disabled && baseStyles.cardDisabled,
      disabled && variantStyles.disabled,
      externalStyles
    ),
    inner: css(disabled && baseStyles.innerDisabled),
  }
}
