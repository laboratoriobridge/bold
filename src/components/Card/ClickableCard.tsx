import React, { ComponentProps } from 'react'
import { ExternalStyles, useCss } from '../../styles'
import { CardVariant, createBaseStyles, createClickableStyles, createVariantStyles } from './CardVariants'

export interface ClickableCardProps extends Omit<ComponentProps<'button'>, 'style'> {
  style?: ExternalStyles

  /**
   * Defines the visual style of the Card
   * @default 'outline'
   */
  variant?: CardVariant

  /**
   * Specify whether the Card should be selected
   */
  selected?: boolean

  /**
   * Specify whether the Card should have a error appearance
   */
  invalid?: boolean

  /**
   * Specify whether the Card should be disabled
   */
  disabled?: boolean
}

export function ClickableCard(props: ClickableCardProps) {
  const { variant = 'outline', children, invalid, disabled, selected, style, ...buttonProps } = props

  const { theme, css } = useCss()

  const baseStyles = createBaseStyles(theme)
  const variantStyles = createVariantStyles(theme)[variant]
  const clickableStyles = createClickableStyles(theme)

  const classes = css(
    baseStyles.card,
    variantStyles.base,
    clickableStyles,
    invalid && variantStyles.invalid,
    disabled && baseStyles.cardDisabled,
    disabled && variantStyles.disabled,
    style
  )

  return (
    <button
      type='button'
      className={classes}
      disabled={disabled}
      data-variant={variant}
      data-selected={selected}
      data-invalid={invalid}
      aria-pressed={selected}
      {...buttonProps}
    >
      <div className={css(disabled && baseStyles.innerDisabled)}>{children}</div>
    </button>
  )
}
