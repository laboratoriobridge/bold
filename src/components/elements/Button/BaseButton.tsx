import * as React from 'react'

export type OnClickWithPromise = (event: React.MouseEvent<any>) => any

export interface BaseButtonProps {
    /**
     * css className
     */
    className?: string
    disabled?: boolean
    name?: string
    onClick?: React.MouseEventHandler<any> | OnClickWithPromise
    onMouseEnter?: React.MouseEventHandler<any>
    onMouseLeave?: React.MouseEventHandler<any>
    tabIndex?: number
    title?: string
    innerRef?: React.Ref<HTMLElement>
    onLoadingChange?(loading: boolean): void
    render?(props: any): React.ReactNode
}

export class BaseButton extends React.Component<BaseButtonProps> {

    static defaultProps: BaseButtonProps = {
        render: ({ innerRef, ...rest }: any) => {
            return (
                <button type='button' ref={innerRef} {...rest} />
            )
        },
    }

    private timeout: number

    componentWillUnmount() {
        clearTimeout(this.timeout)
        this.timeout = -1
    }

    render() {
        const { onLoadingChange, render, ...rest } = this.props

        return render({
            ...rest,
            onClick: this.onClick,
            onKeyPress: this.handleOnKeyPress,
        })
    }

    private handleOnKeyPress = (event) => {
        event.preventDefault()
        if (!event) { event = window.event } // cross-browser shenanigans
        if (event.charCode === 32 || event.charCode === 13 && this.props.onClick) { // this is the spacebar
            this.onClick(event)
        }
        return true // treat all other keys normally;
    }

    private onClick = (event: React.MouseEvent<any>) => {
        if (this.props.onClick) {

            const promise = this.props.onClick(event)
            if (promise && promise.then) {
                this.startLoading()
                promise
                    .then(() => this.stopLoading())
                    .catch((error) => {
                        this.stopLoading()
                        throw new Error(error)
                    })
            }
        }
    }

    private startLoading = () => {
        this.props.onLoadingChange && this.props.onLoadingChange(true)
    }

    private stopLoading() {
        if (!this.timeout) {
            this.timeout = window.setTimeout(() => {
                this.props.onLoadingChange && this.props.onLoadingChange(false)
                this.timeout = null
            }, 10)
        }
    }

}
