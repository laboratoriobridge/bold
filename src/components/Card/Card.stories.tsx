import React from 'react'
import { action } from '@storybook/addon-actions'
import { boolean, select } from '@storybook/addon-knobs'
import { VFlow } from '../VFlow'
import { HeadingSection } from '../HeadingSection'
import { HFlow } from '../HFlow'
import { Checkbox } from '../Checkbox'
import { Button } from '../Button'
import { Icon } from '../Icon'
import { Text } from '../Text'
import { Tag } from '../Tag'
import { Heading } from '../Heading'
import { Status } from '../Status'
import { Card, CardVariant } from './Card'

const variants: { [key in CardVariant]: CardVariant } = {
  outline: 'outline',
  float: 'float',
  plain: 'plain',
}

const Content = () => (
  <HFlow hSpacing={0.5} style={{ gridAutoColumns: 'auto 1fr' }}>
    <Checkbox />
    <HFlow alignItems='center' justifyContent='space-between'>
      <VFlow vSpacing={0.5}>
        <VFlow vSpacing={0.25}>
          <HFlow hSpacing={0.5} alignItems='center'>
            <Heading level={4}>Title</Heading>
            <Tag>Tag label</Tag>
          </HFlow>
          <Text>Description</Text>
        </VFlow>
        <Status type='info' text='Status description' />
      </VFlow>
      <Button skin='ghost' size='small'>
        <Icon icon='adjust' />
      </Button>
    </HFlow>
  </HFlow>
)

export default {
  title: 'Components/Card',
}

export const Default = () => {
  const variant = select('variant', variants, 'outline')
  const selected = boolean('selected', false)
  const invalid = boolean('invalid', false)
  const disabled = boolean('disabled', false)

  return (
    <VFlow>
      <HeadingSection level={2} vSpace={8} title='Interactive'>
        <Card
          variant={variant}
          selected={selected}
          disabled={disabled}
          error={invalid}
          onClick={action('Card clicked')}
        >
          <Content />
        </Card>
      </HeadingSection>
      <HeadingSection level={2} vSpace={8} title='Non-interactive'>
        <Card variant={variant} selected={selected} disabled={disabled} error={invalid}>
          <Content />
        </Card>
      </HeadingSection>
    </VFlow>
  )
}

export const Variants = () => {
  const invalid = boolean('invalid', false)

  return (
    <VFlow>
      <HeadingSection level={2} vSpace={8} title='Outline card'>
        <Card variant='outline' error={invalid}>
          <Content />
        </Card>
      </HeadingSection>
      <HeadingSection level={2} vSpace={8} title='Float card'>
        <Card variant='float' error={invalid}>
          <Content />
        </Card>
      </HeadingSection>
      <HeadingSection level={2} vSpace={8} title='Plain card'>
        <Card variant='plain' error={invalid}>
          <Content />
        </Card>
      </HeadingSection>
    </VFlow>
  )
}

export const Clickable = () => {
  const selected = boolean('selected', false)
  const invalid = boolean('invalid', false)
  const disabled = boolean('disabled', false)

  return (
    <VFlow>
      <HeadingSection level={2} vSpace={8} title='Outline card'>
        <Card
          variant='outline'
          selected={selected}
          disabled={disabled}
          error={invalid}
          onClick={action('Card clicked')}
        >
          <Content />
        </Card>
      </HeadingSection>
      <HeadingSection level={2} vSpace={8} title='Float card'>
        <Card variant='float' selected={selected} disabled={disabled} error={invalid} onClick={action('Card clicked')}>
          <Content />
        </Card>
      </HeadingSection>
      <HeadingSection level={2} vSpace={8} title='Plain card'>
        <Card variant='plain' selected={selected} disabled={disabled} error={invalid} onClick={action('Card clicked')}>
          <Content />
        </Card>
      </HeadingSection>
    </VFlow>
  )
}
