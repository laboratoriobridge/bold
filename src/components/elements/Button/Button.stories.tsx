import { action } from '@storybook/addon-actions'
import { boolean, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import React from 'react'
import { MemoryRouter } from 'react-router'

import { Icon } from '../Icon'
import { Text } from '../textual'

import { Button } from './Button/Button'
import { ButtonKind, ButtonSize, ButtonSkin } from './Button/ButtonSkins'
import { ButtonLink } from './ButtonLink/ButtonLink'

const typeOptions: ButtonKind[] = ['normal', 'primary', 'danger']
const skinOptions: ButtonSkin[] = ['default', 'ghost', 'outline']
const sizeOptions: ButtonSize[] = ['small', 'medium', 'large']

storiesOf('Components|Button', module)
  .add('default', () => (
    <Button
      skin={select('skin', skinOptions, 'default')}
      size={select('size', sizeOptions, 'medium')}
      kind={select('kind', typeOptions, 'normal')}
      loading={boolean('loading', false)}
      disabled={boolean('disabled', false)}
      block={boolean('block', false)}
      onClick={action('button-clicked')}
    >
      {text('children', 'Button')}
    </Button>
  ))
  .add('icon', () => (
    <Button
      skin={select('skin', skinOptions, 'ghost')}
      size={select('size', sizeOptions, 'small')}
      kind={select('kind', typeOptions, 'normal')}
      loading={boolean('loading', false)}
      disabled={boolean('disabled', false)}
      block={boolean('block', false)}
      onClick={action('button-clicked')}
    >
      <Icon icon='adjust' />
    </Button>
  ))
  .add('link', () => (
    <MemoryRouter initialEntries={['/']} initialIndex={0}>
      <ButtonLink
        to='/links'
        skin={select('skin', skinOptions, 'default')}
        size={select('size', sizeOptions, 'medium')}
        kind={select('kind', typeOptions, 'normal')}
        loading={boolean('loading', false)}
        disabled={boolean('disabled', false)}
        block={boolean('block', false)}
        onClick={action('link-clicked')}
      >
        <Icon icon='arrowLeft' style={{ marginRight: '0.5rem' }} />
        <Text>Go back</Text>
      </ButtonLink>
    </MemoryRouter>
  ))
