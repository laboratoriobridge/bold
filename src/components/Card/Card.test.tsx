import React from 'react'
import { render } from '@testing-library/react'
import { Button } from '../Button'
import { Card, CardVariant } from '../Card'

it('should render correctly', () => {
  const { container } = render(<Card>Content</Card>)
  expect(container).toMatchSnapshot()
})

it('should accept "style" prop', () => {
  const { container } = render(<Card style={{ background: 'red' }}>Content</Card>)
  expect(container).toMatchSnapshot()
})

describe('variants', () => {
  it('should render outline variant by default', () => {
    const { getByTestId } = render(<Card data-testid='card-outline'>Content</Card>)
    const card = getByTestId('card-outline')
    expect(card).toHaveStyle('background: rgb(255, 255, 255)')
    expect(card).toHaveStyle('border: 1px solid #d3d4dd')
    expect(getComputedStyle(card).boxShadow).toBe('')
  })

  it('should render float variant styles when specified', () => {
    const { getByTestId } = render(
      <Card variant='float' data-testid='card-float'>
        Content
      </Card>
    )
    const card = getByTestId('card-float')
    expect(card).toHaveStyle('background: rgb(255, 255, 255)')
    expect(card).toHaveStyle('border: 1px solid #d3d4dd')
    expect(card).toHaveStyle(
      'box-shadow: 0 2px 4px -1px rgba(0,0,0,0.2),0 1px 10px 0 rgba(0,0,0,0.12),0 4px 5px 0 rgba(0,0,0,0.14)'
    )
  })

  it('should render plain variant styles when specified', () => {
    const { getByTestId } = render(
      <Card variant='plain' data-testid='card-plain'>
        Content
      </Card>
    )
    const card = getByTestId('card-plain')
    expect(card).toHaveStyle('background: rgb(255, 255, 255)')
    expect(card).toHaveStyle('border: 0px')
    expect(getComputedStyle(card).boxShadow).toBe('')
  })
})

describe('states', () => {
  it.each<CardVariant>(['outline', 'float', 'plain'])(
    'should apply disabled styles in each variant when is disabled',
    (variant) => {
      const { getByTestId } = render(
        <Card variant={variant} disabled data-testid='card-disabled'>
          Content
        </Card>
      )
      const card = getByTestId('card-disabled')
      expect(card).toHaveStyle('cursor: not-allowed')
      expect(card).toHaveStyle('border-color: #d3d4dd')
      expect(card).toHaveStyle('background: rgb(255, 255, 255)')
    }
  )

  it('should apply pointer-events none to children elements when is disabled', () => {
    const { getByTestId } = render(
      <Card disabled>
        <Button data-testid='card-button'>Button</Button>
      </Card>
    )
    const button = getByTestId('card-button')
    expect(button).toHaveStyle('pointer-events: none')
  })

  it('should apply invalid styles in outline variant when error specified', () => {
    const { getByTestId } = render(
      <Card variant='outline' error='Error' data-testid='card-outline-invalid'>
        Content
      </Card>
    )
    const card = getByTestId('card-outline-invalid')
    expect(card).toHaveStyle('border-color: #d01e29')
  })

  it('should apply invalid styles in float variant when error specified', () => {
    const { getByTestId } = render(
      <Card variant='float' error='Error' data-testid='card-float-invalid'>
        Content
      </Card>
    )
    const card = getByTestId('card-float-invalid')
    expect(card).toHaveStyle('border-color: #d01e29')
  })

  it('should apply invalid styles in plain variant when error specified', () => {
    const { getByTestId } = render(
      <Card variant='plain' error='Error' data-testid='card-plain-invalid'>
        Content
      </Card>
    )
    const card = getByTestId('card-plain-invalid')
    expect(getComputedStyle(card).background).toBe('rgb(254, 236, 237)')
  })
})
