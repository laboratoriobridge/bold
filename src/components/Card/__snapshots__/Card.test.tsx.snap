// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept "style" prop 1`] = `
.emotion-1 {
  width: 100%;
  border: 0;
  border-radius: 2px;
  padding: 1rem;
  -webkit-transition: all .2s;
  transition: all .2s;
  background: #FFFFFF;
  border: 1px solid #D3D4DD;
  border-radius: 2px;
  background: red;
}

.emotion-1[data-invalid="true"] {
  border-color: #D01E29;
}

.emotion-1[data-disabled="true"] {
  border-color: #D3D4DD;
  background: #FFFFFF;
}

<div>
  <div
    class="emotion-1"
    data-invalid="false"
  >
    <div
      class="emotion-0"
    >
      Content
    </div>
  </div>
</div>
`;

exports[`should render correctly 1`] = `
.emotion-1 {
  width: 100%;
  border: 0;
  border-radius: 2px;
  padding: 1rem;
  -webkit-transition: all .2s;
  transition: all .2s;
  background: #FFFFFF;
  border: 1px solid #D3D4DD;
  border-radius: 2px;
}

.emotion-1[data-invalid="true"] {
  border-color: #D01E29;
}

.emotion-1[data-disabled="true"] {
  border-color: #D3D4DD;
  background: #FFFFFF;
}

<div>
  <div
    class="emotion-1"
    data-invalid="false"
  >
    <div
      class="emotion-0"
    >
      Content
    </div>
  </div>
</div>
`;
